### ISSUE TITLE
Exception ( Monitor Deploy ) java.lang.NullPointerException

### ISSUE THREAD

**Issue by dpolicastro** [spinnaker/spinnaker#5582]
### Issue Summary:
I'm trying to deploy a K8s Manifests, that is valid, but I'm getting this error:
![image](https://user-images.githubusercontent.com/29067941/76789066-dd153300-679a-11ea-9e1f-c9c00383b4c1.png). I have no ideia of how to troubleshoot this, it started to happen after I upgraded my Spinnaker version from `1.16.x` to `1.19.0`.

### Cloud Provider(s):
AWS/K8s

### Environment:
Spinnaker 1.19.0

### Feature Area:
Clouddriver

### Steps to Reproduce:
- Create application and a pipeline with `Deploy (Manifest)`
- Try to deploy the manifest declared in `additional details` as the `Manifest Source text`

#### K8s Manifests 
```
---
# Source: nodeapp/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: nodeapp
  namespace: nodeapp
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort:
  sessionAffinity: None
  type: ClusterIP
---
# Source: nodeapp/templates/deployment.yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: nodeapp
  namespace: nodeapp
  annotations:
    version: "1"
    strategy.spinnaker.io/max-version-history: '2'
    traffic.spinnaker.io/load-balancers: '["service nodeapp"]'
  labels:
    app: nodeapp
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nodeapp
  template:
    metadata:
      labels:
        app: nodeapp
      annotations:
        timestamp: ""
        version: "1"
    spec:
      containers:
        - name: nodeapp
          image: dpolicastro/node-http-server:latest
          imagePullPolicy: Always
          ports:
          - name: http
            containerPort: 80
            protocol: TCP
          env:
          - name: APP_NAME
            value: nodeapp
          - name: APP_PORT
            value: "80"
          - name: APP_HOST
            value: "0.0.0.0"
          - name: APP_VERSION
            value: "1"
          - name: TIME_BEFORE_START
            value: "5000"
          livenessProbe:
            httpGet:
              path: /health
              port: http
            initialDelaySeconds: 5
            periodSeconds: 5
            successThreshold: 1
            failureThreshold: 3
            timeoutSeconds: 3
          readinessProbe:
            httpGet:
              path: /health
              port: http
            initialDelaySeconds: 5
            periodSeconds: 5
            successThreshold: 1
            failureThreshold: 3
            timeoutSeconds: 3
          resources:
            limits:
              cpu: 100m
              memory: 128Mi
            requests:
              cpu: 100m
              memory: 128Mi
```
#### Clouddriver logs
```
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.455  INFO 1 --- [0.0-7002-exec-2] c.n.s.c.k.v.v.KubernetesValidationUtil   : Validating credentials for default service nodeapp
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.456  INFO 1 --- [0.0-7002-exec-2] c.n.s.c.k.v.v.KubernetesValidationUtil   : Validating credentials for default replicaSet nodeapp
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.473  WARN 1 --- [0.0-7002-exec-2] c.n.s.c.deploy.DescriptionAuthorizer     : No application(s) specified for operation with account restriction (type: KubernetesDeployManifestDescription, rawJson: {"account":"default","credentials":"default","enableTraffic":true,"events":[],"manifests":[{"apiVersion":"v1","kind":"Service","metadata":{"name":"nodeapp","namespace":"nodeapp"},"spec":{"ports":[{"port":80,"protocol":"TCP"}],"sessionAffinity":"None","type":"ClusterIP"}},{"apiVersion":"apps/v1","kind":"ReplicaSet","metadata":{"annotations":{"strategy.spinnaker.io/max-version-history":"2","traffic.spinnaker.io/load-balancers":"[\"service nodeapp\"]","version":"1"},"labels":{"app":"nodeapp"},"name":"nodeapp","namespace":"nodeapp"},"spec":{"replicas":1,"selector":{"matchLabels":{"app":"nodeapp"}},"template":{"metadata":{"annotations":{"timestamp":"","version":"1"},"labels":{"app":"nodeapp"}},"spec":{"containers":[{"env":[{"name":"APP_NAME","value":"nodeapp"},{"name":"APP_PORT","value":"80"},{"name":"APP_HOST","value":"0.0.0.0"},{"name":"APP_VERSION","value":"1"},{"name":"TIME_BEFORE_START","value":"5000"}],"image":"dpolicastro/node-http-server:latest","imagePullPolicy":"Always","livenessProbe":{"failureThreshold":3,"httpGet":{"path":"/health","port":"http"},"initialDelaySeconds":5,"periodSeconds":5,"successThreshold":1,"timeoutSeconds":3},"name":"nodeapp","ports":[{"containerPort":80,"name":"http","protocol":"TCP"}],"readinessProbe":{"failureThreshold":3,"httpGet":{"path":"/health","port":"http"},"initialDelaySeconds":5,"periodSeconds":5,"successThreshold":1,"timeoutSeconds":3},"resources":{"limits":{"cpu":"100m","memory":"128Mi"},"requests":{"cpu":"100m","memory":"128Mi"}}}]}}}}],"moniker":{"app":"test-app"},"namespaceOverride":"","optionalArtifacts":[],"requiredArtifacts":[],"source":"text"})
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.651 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=ORCHESTRATION status=Processing op: KubernetesDeployManifestOperation
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.666 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Beginning deployment of manifest...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.675 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Finding deployer for service...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.681 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Swapping out artifacts in service nodeapp from context...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.682 DEBUG 1 --- [tionProcessor-1] c.n.s.c.k.v2.artifact.ArtifactReplacer   : Doing replacement on {metadata={name=nodeapp, namespace=nodeapp, annotations={}}, apiVersion=v1, kind=Service, spec={ports=[{port=80, protocol=TCP}], sessionAffinity=None, type=ClusterIP}} using []
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.706 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Finding deployer for replicaSet...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.707 DEBUG 1 --- [onPool-worker-1] c.n.s.c.jobs.local.JobExecutorLocal      : Starting job: 'kubectl --namespace=cattle-prometheus-p-dr4gv -o json get configMap,controllerRevision,cronJob,daemonSet,deployment,event,horizontalpodautoscaler,ingress,job,limitRange,networkPolicy,persistentVolumeClaim,pod,podDisruptionBudget,replicaSet,role,roleBinding,secret,service,serviceAccount,statefulSet'...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.712 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Swapping out artifacts in replicaSet nodeapp from context...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.712 DEBUG 1 --- [tionProcessor-1] c.n.s.c.k.v2.artifact.ArtifactReplacer   : Doing replacement on {metadata={annotations={strategy.spinnaker.io/max-version-history=2, traffic.spinnaker.io/load-balancers=["service nodeapp"], version=1}, labels={app=nodeapp}, name=nodeapp, namespace=nodeapp}, apiVersion=apps/v1, kind=ReplicaSet, spec={replicas=1, selector={matchLabels={app=nodeapp}}, template={metadata={annotations={timestamp=, version=1}, labels={app=nodeapp}}, spec={containers=[{env=[{name=APP_NAME, value=nodeapp}, {name=APP_PORT, value=80}, {name=APP_HOST, value=0.0.0.0}, {name=APP_VERSION, value=1}, {name=TIME_BEFORE_START, value=5000}], image=dpolicastro/node-http-server:latest, imagePullPolicy=Always, livenessProbe={failureThreshold=3, httpGet={path=/health, port=http}, initialDelaySeconds=5, periodSeconds=5, successThreshold=1, timeoutSeconds=3}, name=nodeapp, ports=[{containerPort=80, name=http, protocol=TCP}], readinessProbe={failureThreshold=3, httpGet={path=/health, port=http}, initialDelaySeconds=5, periodSeconds=5, successThreshold=1, timeoutSeconds=3}, resources={limits={cpu=100m, memory=128Mi}, requests={cpu=100m, memory=128Mi}}}]}}}} using []
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.724 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Checking if all requested artifacts were bound...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.735 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Sorting manifests by priority...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.741 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Finding deployer for replicaSet...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.748 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Finding deployer for service...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.753 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Deploy order is: service nodeapp, replicaSet nodeapp
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.759 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Finding deployer for service...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.766 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Annotating manifest service nodeapp with artifact, relationships & moniker...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.780 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Swapping out artifacts in service nodeapp from other deployments...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.780 DEBUG 1 --- [tionProcessor-1] c.n.s.c.k.v2.artifact.ArtifactReplacer   : Doing replacement on {metadata={name=nodeapp, namespace=nodeapp, annotations={artifact.spinnaker.io/type=kubernetes/service, artifact.spinnaker.io/name=nodeapp, artifact.spinnaker.io/location=nodeapp, moniker.spinnaker.io/cluster=service nodeapp, moniker.spinnaker.io/application=test-app}, labels={app.kubernetes.io/managed-by=spinnaker, app.kubernetes.io/name=test-app}}, apiVersion=v1, kind=Service, spec={ports=[{port=80, protocol=TCP}], sessionAffinity=None, type=ClusterIP}} using []
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.787 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Submitting manifest service nodeapp to kubernetes master...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.787 DEBUG 1 --- [tionProcessor-1] .v.o.m.KubernetesDeployManifestOperation : Manifest in default to be deployed: {metadata={name=nodeapp, namespace=nodeapp, annotations={artifact.spinnaker.io/type=kubernetes/service, artifact.spinnaker.io/name=nodeapp, artifact.spinnaker.io/location=nodeapp, moniker.spinnaker.io/cluster=service nodeapp, moniker.spinnaker.io/application=test-app}, labels={app.kubernetes.io/managed-by=spinnaker, app.kubernetes.io/name=test-app}}, apiVersion=v1, kind=Service, spec={ports=[{port=80, protocol=TCP}], sessionAffinity=None, type=ClusterIP}}
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:14.789 DEBUG 1 --- [tionProcessor-1] c.n.s.c.jobs.local.JobExecutorLocal      : Starting job: 'kubectl apply -f -'...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:15.196 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Finding deployer for replicaSet...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:15.209 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Annotating manifest replicaSet nodeapp with artifact, relationships & moniker...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:15.212 DEBUG 1 --- [tionProcessor-1] c.n.s.c.jobs.local.JobExecutorLocal      : Starting job: 'kubectl --namespace=nodeapp -o json get service nodeapp'...
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:15.419 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=DEPLOY_KUBERNETES_MANIFEST status=Attaching load balancer service nodeapp to replicaSet nodeapp
[spin-clouddriver-69fdb8b9cd-26zgj] 2020-03-16 18:50:15.451 DEBUG 1 --- [tionProcessor-1] c.n.spinnaker.clouddriver.sql.SqlTask    : Updated status for task 01E3JAX837BTCF8ZWJWM0EVDES phase=ORCHESTRATION status=Orchestration failed: KubernetesDeployManifestOperation | NullPointerException: [java.lang.NullPointerException
[spin-clouddriver-69fdb8b9cd-26zgj] 	at java.util.HashMap.putMapEntries(HashMap.java:501)
[spin-clouddriver-69fdb8b9cd-26zgj] 	at java.util.HashMap.putAll(HashMap.java:785)
[spin-clouddriver-69fdb8b9cd-26zgj] 	at com.netflix.spinnaker.clouddriver.kubernetes.v2.op.handler.KubernetesServiceHandler.attach(KubernetesServiceHandler.java:180)
```

**Comment by maggieneterval** [comment]
@dpolicastro thanks for taking the time to open this issue. One thing I'm noticing off the top of my head is that your Service has no `spec.selector`, which is required for Spinnaker-managed traffic. Could you try adding at least one label selector to your service (different from any labels already associated with your `ReplicaSet`)?

**Comment by dpolicastro** [comment]
Hi @maggieneterval , thanks for helping me figure this out. So, I notice there was a typo in my helm chart letting the `targetPort` key without a value. So I fixed this and added the selector, and the deploy worked! 
```
---
# Source: nodeapp/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: nodeapp
  namespace: nodeapp
spec:
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80 
  selector:
    frontedBy: nodeapp
  sessionAffinity: None
  type: ClusterIP
```

The label being different from the label already used in the RS was a key factor as well, I think its not possible to use the same Key. I tried to use once `app`, and got the following error:

![image](https://user-images.githubusercontent.com/29067941/76817961-18d3eb00-67e3-11ea-90db-3c1f5d9d4803.png)

I tried to replicate it but was not able to, so don't know what happend there. So I think this can be closed. Thanks

**Comment by dpolicastro** [comment]
But before closing it, I was thinking that there should be some kind of Manifest Validation, because the manifest file is valid for Kubernetes (even with the `targetPort: `), but was not valid for Spinnaker to deploy. I spent a lot of time trying to figure out what was the problem, because I first deployed these manifest using `kubectl apply` and everything worked fine.

**Comment by maggieneterval** [comment]
Hey @dpolicastro, thanks for the suggestion: we've added some validation for when a Service is in an invalid state to use Spinnaker managed rollout strategies; these changes will be released with Spinnaker 1.20.
